plutus-tx/src/PlutusTx/Lift/Class.hs

```diff
@@ -59,6 +60,26 @@ import Prettyprinter qualified as PP
 -- We do not use qualified import because the whole module contains off-chain code
 import Prelude as Haskell
 
+#if MIN_VERSION_template_haskell(2,17,0)
```

Comment:
I suggest using https://hackage.haskell.org/package/th-compat for all this stuff.

Reply:
I think that will let you avoid 90% of the CPP.

Reply:
I looked into this, and unfortunately, the majority of the CPP can't really be removed AFAICT.  The only part that can be definitely removed is what you've commented on here. Stuff like this can't be fixed AFAICT:
```haskell
readJSONFromFile :: (FromJSON a, Lift a) => String -> Q (TExp a)
-- to
readJSONFromFile :: (FromJSON a, Lift a) => String -> Code Q a
```

To make splices work as expected, the type needs to change. Don't think there's any way around most of the CPP unfortunately.

Reply:
No, `th-compat` provides stuff to work around that: https://hackage.haskell.org/package/th-compat-0.1.3/docs/Language-Haskell-TH-Syntax-Compat.html#g:5

Reply:
Will `Code q a` still work in splices on template-haskell 2.16?

Reply:
i.e. `$$(readJSONFromFile something)`

Reply:
If you read the bit I linked it explains two ways to have splices that work on both versions, using `IsCode` or `Splice`.

Reply:
I think `IsCode` probably provides what you want.

Reply:
I'll stop being lazy and skim reading.

Reply:
I initially went for `IsCode` but I switched to `Splice` because it was a lot simpler.